### DISCUSSION TITLE
DH group MODP_4096 unacceptable, requesting CURVE_25519

### DISCUSSION THREAD

**Discussion by f1-outsourcing** [strongswan/strongswan#2150] [Q&A]
I have this from android client. I assume the android is requesting MODP_4096 and server is asking for CURVE?

How can I get the server to ask for MODP_2048 (I think I made 2048bit RSA key)

default install strongswan-5.9.12-r0

**Comment by tobiasbrunner** [comment]
> How can I get the server to ask for MODP_2048 (I think I made 2048bit RSA key)

What's negotiated there is the key exchange method from which the IKE and IPsec keys are derived. This has nothing to do with the authentication for which you might be using RSA.

If you want to use specific algorithms, configure the IKE/ESP proposals accordingly (see [the docs](https://docs.strongswan.org/docs/5.9/config/IKEv2CipherSuites.html) for possible options). There is also a global option (`charon.prefer_configured_proposals`) that can make the server prefer the algorithms sent by the clients instead of its own (as long as what clients propose is acceptable to the server).

**Reply by f1-outsourcing** [reply]
https://docs.strongswan.org/docs/5.9/os/androidVpnClient.html
https://docs.strongswan.org/docs/5.9/config/IKEv2CipherSuites.html
https://docs.strongswan.org/docs/5.9/config/IKEv2CipherSuites.html#ike-support

according to these pages, the default win config should be enough, so I guess it is not (with android13) correct? Or is this only related to the strongswan app and not the stock vpn client?

I think android does not load 4096 rsa keys, nor does it seem to handle ec keys. So that is why I had to create 2048 rsa key. But I still don't get what I have to change where.

This is what is logged by the stock vpn client
```
Mar  7 15:58:46 test2 strongswan: 08[NET] received packet: from a.a.a.a[16488] to a.a.a.a[500] (1072 bytes)
Mar  7 15:58:46 test2 strongswan: 08[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) ]
Mar  7 15:58:46 test2 strongswan: 08[IKE] a.a.a.a is initiating an IKE_SA
Mar  7 15:58:46 test2 strongswan: 08[IKE] a.a.a.a is initiating an IKE_SA
Mar  7 15:58:46 test2 strongswan: 08[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519
Mar  7 15:58:46 test2 strongswan: 08[IKE] remote host is behind NAT
Mar  7 15:58:46 test2 strongswan: 08[IKE] DH group MODP_4096 unacceptable, requesting CURVE_25519
Mar  7 15:58:46 test2 strongswan: 08[ENC] generating IKE_SA_INIT response 0 [ N(INVAL_KE) ]
Mar  7 15:58:46 test2 strongswan: 08[NET] sending packet: from a.a.a.a[500] to a.a.a.a[16488] (38 bytes)
Mar  7 15:58:46 test2 strongswan: 15[NET] received packet: from a.a.a.a[16489] to a.a.a.a[500] (592 bytes)
Mar  7 15:58:46 test2 strongswan: 15[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) ]
Mar  7 15:58:46 test2 strongswan: 15[IKE] a.a.a.a is initiating an IKE_SA
Mar  7 15:58:46 test2 strongswan: 15[IKE] a.a.a.a is initiating an IKE_SA
Mar  7 15:58:46 test2 strongswan: 15[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/CURVE_25519
Mar  7 15:58:46 test2 strongswan: 15[IKE] remote host is behind NAT
```

this is logged by the android strongswan vpn app which is connecting

```
Mar  7 15:57:42 test2 strongswan: 16[NET] received packet: from a.a.a.a[16478] to a.a.a.a[500] (948 bytes)
Mar  7 15:57:42 test2 strongswan: 16[ENC] parsed IKE_SA_INIT request 0 [ SA KE No N(NATD_S_IP) N(NATD_D_IP) N(FRAG_SUP) N(HASH_ALG) N(REDIR_SUP) ]
Mar  7 15:57:42 test2 strongswan: 16[IKE] a.a.a.a is initiating an IKE_SA
Mar  7 15:57:42 test2 strongswan: 16[IKE] a.a.a.a is initiating an IKE_SA
Mar  7 15:57:42 test2 strongswan: 16[CFG] selected proposal: IKE:AES_CBC_128/HMAC_SHA2_256_128/PRF_HMAC_SHA2_256/ECP_256
Mar  7 15:57:42 test2 strongswan: 16[IKE] remote host is behind NAT
```

**Reply by tobiasbrunner** [reply]
> according to these pages, the default win config should be enough, so I guess it is not (with android13) correct?

Why not? What's not working? Switching key exchange methods is a normal feature of IKEv2, it's not an error. If you don't like the (EC)DH group that's chosen, change the configuration as I pointed out already.

> I think android does not load 4096 rsa keys, nor does it seem to handle ec keys.

Again, this has absolutely nothing to do with RSA and the authentication. This is related to the IKE proposal, which lists key exchange methods in a specific order. If the server prefers its own config (default), it selects the first proposal and algorithm that's not rejected by the client's proposal. The DH group in the selected proposal then might be different than what the client anticipated. With IKEv2, it has to send a KE payload that contains the public value for a specific DH group in the same message as the proposals, so the group in that KE payload might not match the one proposal the server selected, so the server can request that the client starts again and uses a different group. Again, that's not an error, it's a normal feature of IKEv2.
